CoViewingExample/AudioDevices/ExampleAVPlayerAudioDevice.m

```diff
@@ -0,0 +1,1132 @@
+            NSLog(@"Could not set stream format on the output bus!");
+            AudioComponentInstanceDispose(_voiceProcessingIO);
+            _voiceProcessingIO = NULL;
+            return NO;
+        }
+
+        // Setup the rendering callbacks.
+        UInt32 elementCount = 2;
+        status = AudioUnitSetProperty(_playbackMixer, kAudioUnitProperty_ElementCount,
+                                      kAudioUnitScope_Input, 0, &elementCount,
+                                      sizeof(elementCount));
+        if (status != 0) {
+            NSLog(@"Could not set input element count!");
+            AudioComponentInstanceDispose(_voiceProcessingIO);
+            _voiceProcessingIO = NULL;
+            return NO;
+        }
+
+        AURenderCallbackStruct audioTapRenderCallback;
+        audioTapRenderCallback.inputProc = ExampleAVPlayerAudioDeviceAudioTapPlaybackCallback;
+        audioTapRenderCallback.inputProcRefCon = (void *)(rendererContext);
+        status = AudioUnitSetProperty(_playbackMixer, kAudioUnitProperty_SetRenderCallback,
+                                      kAudioUnitScope_Input, 0, &audioTapRenderCallback,
+                                      sizeof(audioTapRenderCallback));
+        if (status != 0) {
+            NSLog(@"Could not set audio tap rendering callback!");
+            AudioComponentInstanceDispose(_voiceProcessingIO);
+            _voiceProcessingIO = NULL;
+            return NO;
+        }
+
+        AURenderCallbackStruct audioRendererRenderCallback;
+        audioRendererRenderCallback.inputProc = ExampleAVPlayerAudioDeviceAudioRendererPlaybackCallback;
+        audioRendererRenderCallback.inputProcRefCon = (void *)(rendererContext);
+        status = AudioUnitSetProperty(_playbackMixer, kAudioUnitProperty_SetRenderCallback,
+                                      kAudioUnitScope_Input, 1, &audioRendererRenderCallback,
+                                      sizeof(audioRendererRenderCallback));
+        if (status != 0) {
+            NSLog(@"Could not set audio renderer rendering callback!");
+            AudioComponentInstanceDispose(_voiceProcessingIO);
+            _voiceProcessingIO = NULL;
+            return NO;
+        }
+    }
+
+    [self setupAudioCapturer:self.capturingContext];
+
+    // Finally, initialize the IO audio unit and mixer (if present).
+    status = AudioUnitInitialize(_voiceProcessingIO);
+    if (status != noErr) {
+        NSLog(@"Could not initialize the audio unit!");
+        AudioComponentInstanceDispose(_voiceProcessingIO);
+        _voiceProcessingIO = NULL;
+        return NO;
+    }
+
+    if (_playbackMixer) {
+        status = AudioUnitInitialize(_playbackMixer);
+        if (status != noErr) {
+            NSLog(@"Could not initialize the playback mixer audio unit!");
+            AudioComponentInstanceDispose(_playbackMixer);
+            _playbackMixer = NULL;
+            return NO;
+        }
+    }
+
+    return YES;
+}
```

Comment:
can we refactor this method into three parts ?
1. setup audio mixer
2. setup audio renderer
3. setup audio capturer